File: app/lib/crypto/ids.dart
import 'dart:typed_data';

import 'bytes.dart';

final RegExp _uuidRe = RegExp(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$');

/// Parses a canonical UUID string into its 16 bytes.
Uint8List uuidToBytes(String id) {
  final s = id.toLowerCase();
  if (!_uuidRe.hasMatch(s)) throw FormatException('invalid uuid: $id');
  return hexToBytes(s.replaceAll('-', ''));
}

String bytesToUuid(List<int> b) {
  if (b.length != 16) throw const FormatException('uuid must be 16 bytes');
  final h = bytesToHex(b);
  return '${h.substring(0, 8)}-${h.substring(8, 12)}-${h.substring(12, 16)}-${h.substring(16, 20)}-${h.substring(20)}';
}

/// Random UUID version 4.
String newUuid() {
  final b = randomBytes(16);
  b[6] = (b[6] & 0x0f) | 0x40;
  b[8] = (b[8] & 0x3f) | 0x80;
  return bytesToUuid(b);
}

File: web/src/crypto/ids.ts
import { bytesToHex, hexToBytes } from './bytes';

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/** Parses a canonical UUID string into its 16 bytes. */
export function uuidToBytes(id: string): Uint8Array {
  const s = id.toLowerCase();
  if (!UUID_RE.test(s)) throw new Error(`invalid uuid: ${id}`);
  return hexToBytes(s.replace(/-/g, ''));
}

export function bytesToUuid(b: Uint8Array): string {
  if (b.length !== 16) throw new Error('uuid must be 16 bytes');
  const h = bytesToHex(b);
  return `${h.slice(0, 8)}-${h.slice(8, 12)}-${h.slice(12, 16)}-${h.slice(16, 20)}-${h.slice(20)}`;
}

export function isUuid(s: string): boolean {
  return UUID_RE.test(s.toLowerCase());
}

export function newUuid(): string {
  return crypto.randomUUID();
}
